daemon/api_validate.go

```diff
@@ -0,0 +1,260 @@
+		return BadRequest("invalid account ID %q", accountID)
+	}
+	if !asserts.IsValidValidationSetName(name) {
+		return BadRequest("invalid name %q", name)
+	}
+
+	query := r.URL.Query()
+
+	// sequence is optional
+	sequenceStr := query.Get("sequence")
+	var sequence int
+	if sequenceStr != "" {
+		var err error
+		sequence, err = strconv.Atoi(sequenceStr)
+		if err != nil {
+			return BadRequest("invalid sequence argument")
+		}
+		if sequence < 0 {
+			return BadRequest("invalid sequence argument: %d", sequence)
+		}
+	}
+
+	st := c.d.overlord.State()
+	st.Lock()
+	defer st.Unlock()
+
+	var tr assertstate.ValidationSetTracking
+	err := assertstate.GetValidationSet(st, accountID, name, &tr)
+	if err == state.ErrNoState || (err == nil && sequence != 0 && sequence != tr.PinnedAt) {
+		return validationSetNotFound(accountID, name, sequence)
+	}
+	if err != nil {
+		return InternalError("accessing validation sets failed: %v", err)
+	}
+
+	modeStr, err := modeString(tr.Mode)
+	if err != nil {
+		return InternalError(err.Error())
+	}
+	// TODO: evaluate against installed snaps
+	var valid bool
+	res := validationSetResult{
+		ValidationSet: validationSetKeyString(tr.AccountID, tr.Name, tr.PinnedAt),
+		Mode:          modeStr,
+		Seq:           tr.Current,
+		Valid:         valid,
+	}
+	return SyncResponse(res, nil)
+}
+
+type validationSetApplyRequest struct {
+	Action   string `json:"action"`
+	Mode     string `json:"mode"`
+	Sequence int    `json:"sequence,omitempty"`
+}
+
+func applyValidationSet(c *Command, r *http.Request, _ *auth.UserState) Response {
+	vars := muxVars(r)
+	accountID := vars["account"]
+	name := vars["name"]
+
+	if !asserts.IsValidAccountID(accountID) {
+		return BadRequest("invalid account ID %q", accountID)
+	}
+	if !asserts.IsValidValidationSetName(name) {
+		return BadRequest("invalid name %q", name)
+	}
+
+	var req validationSetApplyRequest
+	decoder := json.NewDecoder(r.Body)
+	if err := decoder.Decode(&req); err != nil {
+		return BadRequest("cannot decode request body into validation set action: %v", err)
+	}
+	if decoder.More() {
+		return BadRequest("extra content found in request body")
+	}
+	if req.Sequence < 0 {
+		return BadRequest("invalid sequence argument: %d", req.Sequence)
+	}
+
+	st := c.d.overlord.State()
+	st.Lock()
+	defer st.Unlock()
+
+	if req.Action == "forget" {
+		return forgetValidationSet(st, accountID, name, req.Sequence)
+	}
+	if req.Action != "apply" {
+		return BadRequest("unsupported action %q", req.Action)
+	}
+
+	var mode assertstate.ValidationSetMode
+	switch req.Mode {
+	case "monitor":
+		mode = assertstate.Monitor
+	case "enforce":
+		mode = assertstate.Enforce
+	default:
+		return BadRequest("invalid mode %q", req.Mode)
+	}
+
+	// TODO: check that it exists in the store?
+
+	tr := assertstate.ValidationSetTracking{
+		AccountID: accountID,
+		Name:      name,
+		Mode:      mode,
+		// note, Sequence may be 0, meaning not pinned.
+		PinnedAt: req.Sequence,
+	}
```

Comment:
if the mode is enforced here we need to check that the assertion is valid before saving it, for one that comes from the store we would also add it to the assertion db,  we might need more helpers in assertstate for this